### PR TITLE
Support per-block bundling for static memory pools.

### PR DESCRIPTION
**Pull request by doru1004**
This patch ensures that static memory pools get aggregated on a per-block basis.

This also fixes issue https://github.com/onnx/onnx-mlir/issues/323

### CODE REVIEW

File: src/Transform/BundleMemoryPools.cpp [hunk 1]
```diff
@@ -124,6 +128,9 @@ bool isBlockArgument(Block *block, Value operand) {
 KrnlGetRefOp getUnbundledGetRef(AllocOp *memPool) {
   auto parentBlock = memPool->getOperation()->getBlock();
 
+  parentBlock->dump();
+  printf("\n\n");
```
Comment (AlexandreEichenberger) [inline src/Transform/BundleMemoryPools.cpp:132]: may need to remove printfs?

**Review by AlexandreEichenberger — COMMENTED** [review]
When you have a branch

```
if () A else B
```

and you add a block for handling allocs in A, do you have to change the control flow jumping from the test to A-alloc instead of A?

Also, the test that the German folks were looking into

```
block A: %1 = alloc()
if ? goto B
bloc B: %2 = alloc; free %1, return %2
block C %3 = merge (%1 on else (skip block B) path, %2 on then (through block B) path)
free %3.
```
That is a case where the allocs are escaping their blocks...

Can you just confirm that you can safely "not handle" such cases. Thanks

**Comment by doru1004** [comment]
> When you have a branch
> 
> ```
> if () A else B
> ```
> 
> and you add a block for handling allocs in A, do you have to change the control flow jumping from the test to A-alloc instead of A?
> 
> Also, the test that the German folks were looking into
> 
> ```
> block A: %1 = alloc()
> if ? goto B
> bloc B: %2 = alloc; free %1, return %2
> block C %3 = merge (%1 on else (skip block B) path, %2 on then (through block B) path)
> free %3.
> ```
> 
> That is a case where the allocs are escaping their blocks...
> 
> Can you just confirm that you can safely "not handle" such cases. Thanks

Such cases are outside the scope of the bundling algorithm for now. Moreover we rely on MLIR for inserting the deallocs so if there is an error due to wrongly introduced dealloc that means the support in MLIR is not sufficient (yet). There is work by the community to improve this aspect of MLIR.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, thanks also for the big new test

**Pull request merged**